compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -764,20 +764,37 @@ class ExpressionCodegen(
 
     override fun <T> visitConst(expression: IrConst<T>, data: BlockInfo): PromisedValue {
         expression.markLineNumber(startOffset = true)
-        when (val value = expression.value) {
-            is Boolean -> {
-                // BooleanConstants _may not_ be materialized, so we ensure an instruction for the line number.
-                mv.nop()
-                return BooleanConstant(this, value)
+        val value = expression.value
+        val type = when (value) {
+            is Boolean -> return BooleanConstant(this, value)
+            is Byte -> Type.BYTE_TYPE
+            is Char -> Type.CHAR_TYPE
+            is Long -> Type.LONG_TYPE
+            is Short -> Type.SHORT_TYPE
+            is Float -> Type.FLOAT_TYPE
+            is Double -> Type.DOUBLE_TYPE
+            is Number -> Type.INT_TYPE
+            is String -> AsmTypes.JAVA_STRING_TYPE
+            else -> OBJECT_TYPE
+        }
+        return object : PromisedValue(this, type, expression.type) {
+            override fun materializeAt(target: Type, irTarget: IrType) {
+                when (value) {
+                    is Char -> mv.iconst(value.toInt())
+                    is Long -> mv.lconst(value)
+                    is Float -> mv.fconst(value)
+                    is Double -> mv.dconst(value)
+                    is Number -> mv.iconst(value.toInt())
+                    else -> mv.aconst(value)
+                }
+                // `null` is of any reference type.
+                if (value != null || isPrimitive(target)) {
```

Comment:
Why non-primitive types are not cast?

Reply:
Because `aconst null` pushes a value of type "bottom", which is not representable as an asm type but is a subtype of any reference type and does not need casting.

Reply:
Yes, sorry for stupid question. 
But `isPrimitive(target)` check looks redundant, no?

Reply:
Probably, unboxing null into a primitive does not sound like a valid operation.

Reply:
Yes, thank you!

Reply:
Looks like this change exposes additional problems with types in Fir,
there are at least 3 failing test in running build:
FirBlackBoxCodegenTestGenerated$Classes.testKt1018
FirBytecodeTextTestGenerated$NullCheckOptimization.testRedundantSafeCall_1_4
FirBytecodeTextTestGenerated$NullCheckOptimization.testRedundantSafeCall

Reply:
There is also fail with `IrKotlinSteppingTestGenerated$Custom.testConstantConditions`

Reply:
`RedundantSafeCall`s are unmuted by the PR, some recent commit must've broken them in another way.

Reply:
Oh, right, those tests have precisely the problem that null is unboxed into a primitive - on an unreachable code path.